packages/neon-api/src/notifications/class.ts

```diff
@@ -0,0 +1,99 @@
+import { logging, settings } from "@cityofzion/neon-core";
+import WebSocket from "isomorphic-ws";
+import { NotificationMessage, CallbackFunction } from "./responses";
+import Subscription from "./subscription";
+const log = logging.default("api");
+
+interface ContractSubscriptions {
+  websocket: WebSocket;
+  callbacks: CallbackFunction[];
+}
+
+export class Notifications {
+  private url: string;
+
+  public get name(): string {
+    return `Notifications[${this.url}]`;
+  }
+
+  private subscriptions: Map<string | null, ContractSubscriptions>;
+
+  /**
+   * Create a new notification service that handles contract subscriptions
+   * @param url - URL of a notifications service.
+   */
+  public constructor(url: string) {
+    this.url = settings.networks[url]?.extra?.notifications ?? url;
+    this.subscriptions = new Map<string | null, ContractSubscriptions>();
+    log.info(`Created Notifications Provider: ${this.url}`);
+  }
+
+  /**
+   * Subscribe to event notifications of a specific contract
+   * @param contract - Hash of the contract (null for all contracts) to subscribe to
+   * @param callback - Function to call when a notification is received.
+   * @return Subscription object that can be used to cancel the subscription
+   */
+  public subscribe(
+    contract: string | null,
+    callback: CallbackFunction
+  ): Subscription {
+    if (this.subscriptions.has(contract)) {

```

Comment:
Optional, a cleaner flow would be to first pull out the `ContractSubscriptions` object first and just do a push. 

```
const contractSubscriptions = this.subscriptions.get(contract) ?? createWebsocketForContract(contract);
contractSubscriptions.callbacks.push(callback);
```

Reply:
Done in 50387ee49184e0db7caf59e2995ee40c5ad183c3